### PR TITLE
Handled possible null value in predicated entry-processors

### PR DESCRIPTION
**Pull request by ahmetmircik**
Things done:
-  Predicated entry processors should only process entry only if it exists in map.

fixes https://github.com/hazelcast/hazelcast/issues/11452

no need backport, only happens in master

### CODE REVIEW

**Comment by ahmetmircik** [comment]
run-lab-run

**Comment by ahmetmircik** [comment]
run-lab-run

**Comment by ahmetmircik** [comment]
run-lab-run

**Comment by ahmetmircik** [comment]
run-lab-run

**Comment by mmedenjak** [comment]
run-lab-run

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/PartitionWideEntryOperation.java [hunk 1]
```diff
@@ -76,12 +76,18 @@ public void run() {
         responses = new MapEntries(recordStore.size());
         operator = operator(this, entryProcessor, getPredicate(), true);
 
-        Iterator<Record> iterator = recordStore.iterator(Clock.currentTimeMillis(), false);
+        Iterator<Record> iterator = recordStore.loadAwareIterator(Clock.currentTimeMillis(), true, false);
```
Comment (mmedenjak) [inline hazelcast/src/main/java/com/hazelcast/map/impl/operation/PartitionWideEntryOperation.java]: The javadoc of the class needs to be updated. 
If I understand correctly, before the record store was checked if it was loaded and entry access stats were updated on each call to `DefaultRecordStore#get` and now we moved this logic to the iterator. 
Is this correct? Why did it not load the entries before? What do we achieve with this?

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/PartitionWideEntryBackupOperation.java [hunk 2]
```diff
@@ -44,10 +45,16 @@ public void run() {
         responses = new MapEntries(recordStore.size());
         EntryOperator operator = operator(this, backupProcessor, getPredicate(), true);
 
-        Iterator<Record> iterator = recordStore.iterator(Clock.currentTimeMillis(), true);
+        Iterator<Record> iterator = recordStore.loadAwareIterator(Clock.currentTimeMillis(), true, true);
```
Comment (mmedenjak) [inline hazelcast/src/main/java/com/hazelcast/map/impl/operation/PartitionWideEntryBackupOperation.java]: This will actually make this the only 2 entry operations that don't call the get interceptors before getting the values.
Also, if we choose to use the values from the iterator, I guess the HD partition wide operations might be changed as well.

**Comment by jerrinot** [comment]
@ahmetmircik: what's the state of this PR? is it just a regression test missing?

**Comment by ahmetmircik** [comment]
@jerrinot correct.

**Comment by mmedenjak** [comment]
@ahmetmircik can you add the test? Do we know why the value was `null` in the issue in the first place? Was there some expiration? Or did the offloading EP remove the entry? Do we need to add the same check in other EP operations (`HDPartitionWideEntryOperation`, `EntryOperation`, `HDEntryOperation`, `MultipleEntryOperation`, `HDMultipleEntryOperation`) as they might suffer from the same issue?

**Comment by ahmetmircik** [comment]
@jerrinot @mmedenjak 
I added tests, will create backport and ee-ports after reviews are ok.

**Comment by mmedenjak** [comment]
Do you have any info on these?
- Do we know why the value was null in the issue in the first place? Was there some expiration? Or did the offloading EP remove the entry? 
- Do we need to add the same check in other EP operations (HDPartitionWideEntryOperation, EntryOperation, HDEntryOperation, MultipleEntryOperation, HDMultipleEntryOperation) as they might suffer from the same issue?

**Comment by ahmetmircik** [comment]
@mmedenjak 
>Do we know why the value was null in the issue in the first place? Was there some expiration? Or did the offloading EP remove the entry?

One cause of failure can be expiration. I don't have any other scenario related to off-loadables for now. In anyway this PR can be seen a step forward.

>Do we need to add the same check in other EP operations (HDPartitionWideEntryOperation, EntryOperation, HDEntryOperation, MultipleEntryOperation, HDMultipleEntryOperation) as they might suffer from the same issue?

No, the fix here is sufficient for all, same code is used for all.

**Comment by tombujok** [comment]
@ahmetmircik Danny's test does not have expirations, so it is not this one. It's tricky to me why the value may be null. We can't merge it without fully understanding why the value is null. Could please you try to reproduce it to fully understand what happens there? It has a high priority

**Comment by ahmetmircik** [comment]
@tombujok @mmedenjak 
A possible scenario that results with NPE can be like this:
1. split brain happened
2. cluster merge started
3. cluster merge reset record-store data on joining side
4. during all this healing process we have a long running partition operation and it is still running during cluster-merge
5. and it sees some values inside recordstore null as a result of record-store reset at step 3.

According the logs of above run, NPE is happening on joining node after [LocalMemberResetException](https://github.com/hazelcast/hazelcast/blob/7738c39e55148954abc92e228251b76ffe2b7f12/hazelcast/src/main/java/com/hazelcast/core/LocalMemberResetException.java#L23).

**Comment by mmedenjak** [comment]
@ahmetmircik sounds good to me. Is it possible to make a test with this scenario where you start a long-running EP, split and then merge?

**Comment by tombujok** [comment]
Seems legit.

**Comment by ahmetmircik** [comment]
run-lab-run
